Source/Core/DolphinQt/GameList/GameTracker.cpp

```diff
@@ -46,92 +48,76 @@ GameTracker::GameTracker(QObject* parent) : QFileSystemWatcher(parent)
   });
 
   connect(&Settings::Instance(), &Settings::MetadataRefreshRequested, [this] {
-    m_load_thread.EmplaceItem(Command{CommandType::UpdateMetadata, {}});
+    m_cache.UpdateAdditionalMetadata(
+        [this](const std::shared_ptr<const UICommon::GameFile>& game) { emit GameUpdated(game); },
+        m_aborted);
+    Settings::Instance().NotifyMetadataRefreshComplete();
   });
 
-  m_load_thread.Reset([this](Command command) {
-    switch (command.type)
-    {
-    case CommandType::LoadCache:
-      LoadCache();
-      break;
-    case CommandType::Start:
-      StartInternal();
-      break;
-    case CommandType::AddDirectory:
-      AddDirectoryInternal(command.path);
-      break;
-    case CommandType::RemoveDirectory:
-      RemoveDirectoryInternal(command.path);
-      break;
-    case CommandType::UpdateDirectory:
-      UpdateDirectoryInternal(command.path);
-      break;
-    case CommandType::UpdateFile:
-      UpdateFileInternal(command.path);
-      break;
-    case CommandType::UpdateMetadata:
-      m_cache.UpdateAdditionalMetadata(
-          [this](const std::shared_ptr<const UICommon::GameFile>& game) {
-            emit GameUpdated(game);
-          });
-      QueueOnObject(this, [] { Settings::Instance().NotifyMetadataRefreshComplete(); });
-      break;
-    case CommandType::PurgeCache:
-      m_cache.Clear(UICommon::GameFileCache::DeleteOnDisk::Yes);
-      break;
-    case CommandType::BeginRefresh:
-      if (m_busy_count++ == 0)
-      {
-        for (auto& file : m_tracked_files.keys())
-          emit GameRemoved(file.toStdString());
-        m_tracked_files.clear();
-      }
-      break;
-    case CommandType::EndRefresh:
-      if (--m_busy_count == 0)
-      {
-        QueueOnObject(this, [] { Settings::Instance().NotifyRefreshGameListComplete(); });
-      }
-      break;
-    }
-  });
+  QThread::start(QThread::IdlePriority);
 
-  m_load_thread.EmplaceItem(Command{CommandType::LoadCache, {}});
+  LoadCache();
 
   // TODO: When language changes, reload m_title_database and call m_cache.UpdateAdditionalMetadata
 }
 
-void GameTracker::LoadCache()
+GameTracker::~GameTracker()
 {
-  m_cache.Load();
-  m_cache_loaded_event.Set();
+  m_aborted = true;
+  QThread::quit();
+  const bool joined = QThread::wait(5000);
+  assert(joined && "Thread is taking too long to finish");
+  if (!joined)
+  {
+    QThread::terminate();
+  }
 }
 
 void GameTracker::Start()
 {
-  if (m_initial_games_emitted)
-    return;
+  QMetaObject::invokeMethod(this, "OnStart");
+}
+
+void GameTracker::AddDirectory(const QString& dir)
+{
+  QMetaObject::invokeMethod(this, "OnAddDirectory", Q_ARG(QString, dir));
```

Comment:
Instead of these `invokeMethod` calls couldn't you use signals/slots?
That would be significantly cleaner.

Reply:
I have now added a number of signals, connected to the slots, that are emitted when the public functions are invoked, guaranteeing that the slots run in the GameTracker's thread.

Basically:

- `Start()` -> `emit StartRequested()` -> `OnStartRequested()`
- `AddDirectory()` -> `emit AddDirectoryRequested()` -> `OnAddDirectoryRequested()`
- `RemoveDirectory()` -> `emit RemoveDirectoryRequested()` -> `OnRemoveDirectoryRequested()`
- `RefreshAll()` -> `emit RefreshAllRequested()` -> `OnRefreshAlltRequested()`
- `PurgeCache()` -> `emit PurgeCacheRequested()` -> `OnPurgeCacheRequested()`
- `LoadCache()` -> `emit LoadCacheRequested()` -> `OnLoadCacheRequested()`

The only reason why I was using `QMetaObject::invokeMethod()` was to avoid having to declare/connect the signals, but I agree your way is _significantly_ cleaner.